src/core/thread/mesh_forwarder.cpp

```diff
@@ -2271,4 +2279,75 @@ void MeshForwarder::HandleDataPollTimeout(void)
     }
 }
 
+#if OPENTHREAD_CONFIG_LOG_LEVEL >= OPENTHREAD_LOG_LEVEL_INFO
+
+void MeshForwarder::LogIp6Message(MessageAction aAction, const Message &aMessage, const Mac::Address &aMacAddress,
+                                  ThreadError aError) const
+{
+    uint16_t checksum = 0;
+    Ip6::Header ip6Header;
+    Ip6::IpProto protocol;
+    char stringBuffer[(Ip6::Address::kIp6AddressStringSize > Mac::Address::kAddressStringSize) ?
+                      Ip6::Address::kIp6AddressStringSize : Mac::Address::kAddressStringSize];
+
+    VerifyOrExit(aMessage.GetType() == Message::kTypeIp6, ;);
+
+    VerifyOrExit(sizeof(ip6Header) == aMessage.Read(0, sizeof(ip6Header), &ip6Header), ;);
+    VerifyOrExit(ip6Header.IsVersion6(), ;);
+
+    protocol = ip6Header.GetNextHeader();
+
+    switch (protocol)
+    {
+    case Ip6::kProtoUdp:
+    {
+        Ip6::UdpHeader udpHeader;
+
+        VerifyOrExit(sizeof(udpHeader) == aMessage.Read(sizeof(ip6Header), sizeof(udpHeader), &udpHeader), ;);
+        checksum = udpHeader.GetChecksum();
+        break;
+    }
+
+    case Ip6::kProtoTcp:
+    {
+        Ip6::TcpHeader tcpHeader;
+
+        VerifyOrExit(sizeof(tcpHeader) == aMessage.Read(sizeof(ip6Header), sizeof(tcpHeader), &tcpHeader), ;);
+        checksum = tcpHeader.GetChecksum();
+        break;
+    }
+
+    default:
+        break;
+    }
+
+    otLogInfoMac(
+        GetInstance(),
+        "%s IPv6 %s msg, len:%d, chksum:%04x, %s:%s, sec:%s%s%s",
+        (aAction == kMessageReceive) ? "Rx" : ((aError == kThreadError_None) ? "Tx" : "Failed to Tx"),
+        Ip6::Ip6::IpProtoToString(protocol),
+        aMessage.GetLength(),
+        checksum,
+        (aAction == kMessageReceive) ? "from" : "to",
+        aMacAddress.ToString(stringBuffer, sizeof(stringBuffer)),
+        aMessage.IsLinkSecurityEnabled() ? "yes" : "no",
+        (aError == kThreadError_None) ? "" : ", error:",
+        (aError == kThreadError_None) ? "" : otThreadErrorToString(aError)
+    );
+
+    otLogInfoMac(GetInstance(), "src: %s", ip6Header.GetSource().ToString(stringBuffer, sizeof(stringBuffer)));
+    otLogInfoMac(GetInstance(), "dst: %s", ip6Header.GetDestination().ToString(stringBuffer, sizeof(stringBuffer)));
```

Comment:
My main concern is that if you split related info between multiple logs, on Windows, the information can get easily separated because of multiple threads logging simultaneously. If there is a platform limitation on the length of one log message, should it be the platform's responsibility to split into multiple messages?